app/lib/common/widgets/room/room_hierarchy_join_button.dart

```diff
@@ -0,0 +1,81 @@
+import 'package:acter/common/providers/room_providers.dart';
+import 'package:acter/common/utils/rooms.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+
+class RoomHierarchyJoinButton extends ConsumerWidget {
+  final Function(String) forward;
+  final String roomId;
+  final String joinRule;
+  final String roomName;
+  final String? viaServerName;
+
+  const RoomHierarchyJoinButton({
+    super.key,
+    required this.joinRule,
+    required this.roomId,
+    required this.roomName,
+    required this.viaServerName,
+    required this.forward,
+  });
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final maybeRoom = ref.watch(maybeRoomProvider(roomId)).valueOrNull;
+    if (maybeRoom != null) {
+      // we know that room already \o/
+      return OutlinedButton(
+        onPressed: () => forward(roomId),
+        child: Text(L10n.of(context).joined),
+      );
+    }
+    switch (joinRule) {
+      case 'private':
+      case 'invite':
+        return Tooltip(
+          message: L10n.of(context).youNeedBeInvitedToJoinThisRoom,
+          child: Chip(label: Text(L10n.of(context).private)),
+        );
+      case 'restricted':
```

Comment:
So this `restricted` is members only permission where you need to be invited.

We try to call Join API and it will return error if user is not invited, correct?

Reply:
> So this restricted is members only permission where you need to be invited.

no. restricted and you can see it means you can join it.